kore/src/Kore/Step/Simplification/Not.hs

```diff
@@ -110,14 +116,14 @@ simplifyEvaluated
     termSimplifier
     axiomSimplifiers
     simplified
-  | OrPattern.isFalse simplified = return (OrPattern.fromPatterns [Pattern.top])
-  | OrPattern.isTrue  simplified = return (OrPattern.fromPatterns [])
-  | otherwise =
-    fmap OrPattern.fromPatterns . gather
-    $ Foldable.foldrM mkAnd Pattern.top (simplified >>= makeEvaluate)
+  =
+    fmap OrPattern.fromPatterns $ gather $ do
+        let not' = Not { notChild = simplified, notSort = () }
+        andPattern <- scatterAnd (evaluateNotPattern <$> distributeNot not')
```

Comment:
This is reasonable, but you seem to want to change the implicit conventions for the simplification code. To be more precise, usually the simplification code unwraps the various data structures, solves some of the issues that should be solving at that unwrapping level, and sends whatever is left to lower lever simplification stuff (e.g. `simplify` takes a `Not(MultiOr)`, sends the children to `simplifyEvaluated`, which does whatever is needed at the `MultiOr` level (i.e. `not(or)=and(not)`, sending the individual patterns inside the `or` to `makeEvaluate` to negate them, then `and`ing the results to produce the final thing).

Now, it seems that you prefer to work with unwrapped stuff. I think this is reasonable in itself, but I wanted to remind you that the unification code becomes less uniform this way. Up to you to fix or ignore.

Reply:
I acknowledge that one downside of this is that the unification code becomes less uniform. However, I really messed up this simplifier the last time I worked on it, and I think the wrappers (or, the types of the wrappers) are a useful guide. Specifically, I messed up the order in which I distributed the `\and` over the `\or` and applied de Morgan's identity, and with the types of the wrappers in place, it becomes much more difficult to do that.